### PR TITLE
Treat cells to be h-coarsened as their parent in `hp::Refinement::limit_p_level_difference()`.

### PR DESCRIPTION
**Pull request by marcfehling**
Follow-up to #11937.

Re-opening of #11929. See PR description for details.

### CODE REVIEW

File: include/deal.II/hp/refinement.h [hunk 1]
```diff
@@ -677,6 +677,13 @@ namespace hp
      * to call this function, nor will it be automatically invoked in any part
      * of the library (contrary to its Triangulation counterpart).
      *
+     * On cells that will be h-coarsened, we enforce the difference criterion as
+     * if it's already a parent cell. That means, we set the level of all
```
Comment (peterrum) [inline include/deal.II/hp/refinement.h:681]:
```suggestion
     * if it is already a parent cell. That means, we set the level of all
```

File: source/hp/refinement.cc [hunk 2]
```diff
@@ -843,6 +843,97 @@ namespace hp
...
+      // This function takes a cell and sets all its siblings to the highest
+      // p-level among them. Returns whether any future levels have been
+      // changed.
+      const auto prepare_level_for_parent = [&](const auto &neighbor) -> bool {
+        Assert(neighbor->is_active(), ExcInternalError());
+        if (neighbor->coarsen_flag_set() && neighbor->is_locally_owned())
+          {
+            const auto parent = neighbor->parent();
+
+            std::set<unsigned int> future_levels_children;
```
Comment (peterrum) [inline source/hp/refinement.cc:897]: Why not use here a vector?

**Review by peterrum — APPROVED** [review]

File: source/hp/refinement.cc (hunk 2 above)
Comment (marcfehling) [reply source/hp/refinement.cc:897]: I access the last element of the `set` to determine the highest value.
I could also use `vector` and use `max_element` instead. I don't know what's more practical. Any opinions?

Comment (peterrum) [reply source/hp/refinement.cc:897]: Vector :)

**Review by peterrum — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]

**Pull request merged**